src/NCPkgFilterService.cc

```diff
@@ -0,0 +1,322 @@
+|                        |_|\__,_|____/ |_| |_____|                    |
+|                                                                      |
+|                               core system                            |
+|                                                        (C) SuSE GmbH |
+\----------------------------------------------------------------------/
+
+   File:       NCPkgFilterService.cc
+
+/-*/
+#define YUILogComponent "ncurses-pkg"
+#include <YUILog.h>
+
+#include <zypp/ServiceInfo.h>
+
+#include "NCPkgFilterService.h"
+
+#include "YDialog.h"
+#include "NCLayoutBox.h"
+#include "NCSpacing.h"
+#include "NCPackageSelector.h"
+
+#include "NCZypp.h"
+
+using std::endl;
+
+/*
+  Textdomain "ncurses-pkg"
+*/
+
+///////////////////////////////////////////////////////////////////
+//
+//
+//	METHOD NAME : NCPkgServiceTag::NCPkgServiceTag
+//	METHOD TYPE : Constructor
+//
+//	DESCRIPTION :
+//
+
+NCPkgServiceTag::NCPkgServiceTag ( ZyppService servicePtr)
+    : YTableCell(std::string(" "))
+    , service (servicePtr)
+{
+
+}
+
+///////////////////////////////////////////////////////////////////
+//
+//
+//	METHOD NAME : NCPkgServiceTable::NCPkgServiceTable
+//	METHOD TYPE : Constructor
+//
+//	DESCRIPTION :
+//
+
+NCPkgServiceTable::NCPkgServiceTable( YWidget *parent, YTableHeader *tableHeader, NCPackageSelector *pkg )
+    :NCTable( parent, tableHeader )
+    ,packager(pkg)
+    ,repo_manager(new zypp::RepoManager())
+{
+   fillHeader();
+   fillServiceList();
+}
+
+bool NCPkgServiceTable::any_service()
+{
+    bool ret = std::any_of(ZyppRepositoriesBegin(), ZyppRepositoriesEnd(), [](const zypp::Repository& repo) {
+        // if the repository does not belong to any service then the service name is empty
+        return !repo.info().service().empty();
+    });
+
+    yuiMilestone() << "Found a libzypp service: " << ret << std::endl;
+    return ret;
+}
+
+///////////////////////////////////////////////////////////////////
+//
+//
+//	METHOD NAME : NCPkgServiceTable::fillHeader
+//	METHOD TYPE : void
+//
+//	DESCRIPTION : Fill header of servicesitories table (name + URL)
+//
+
+void NCPkgServiceTable::fillHeader()
+{
+    std::vector <std::string> header;
+
+    header.reserve(2);
+    header.push_back( "L" );
+    header.push_back( "L" + NCPkgStrings::PkgName() );
+
+    setHeader( header);
+}
+
+///////////////////////////////////////////////////////////////////
+//
+//
+//	METHOD NAME : NCPkgServiceTable::addLine
+//	METHOD TYPE : void
+//
+//	DESCRIPTION : Add one line (with tag) to the service table
+//
+
+void NCPkgServiceTable::addLine ( ZyppService r, const std::vector <std::string> & cols )
+{
+    //use default ctor, add cell in the next step
+    YTableItem *tabItem = new YTableItem();
+
+    //place tag (with service reference) to the 0th column
+    tabItem->addCell( new NCPkgServiceTag ( r ) );
+
+    // and append the rest (name, URL and stuff)
+    for ( unsigned i = 1; i < cols.size() + 1; ++i ) {
```

Comment:
IMHO it's unintuitive to iterate from 1 to size + 1 and then use i-1 as the index each time. Why not simply iterate (as usual in C/C++) from 0 to size()-1 and use just i as the index?

Just a suggestion.

Reply:
I'd avoid the plain `for` whenever possible, all that `+1`, `-1` logic is fragile. As it is `std::vector` we can use the C++ way easily:

`std::for_each(++cols.begin(), cols.end(), ....)`

Reply:
OMG you're right, the original code is just a mutation of a perfectly ordinary iteration. I am fixing also the other copies.